scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,256 @@
+            // so we have to create a checkpoint with lower timestamp.
+            if checkpoint.0 > Duration::from_millis(1) {
+                checkpoint.0 -= Duration::from_millis(1);
+                checkpoint_maker.make_checkpoint(checkpoint).await?;
+            }
+
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for creating checkpoints.
+#[async_trait]
+pub trait CDCCheckpointMaker: Send + Sync {
+    /// Creates new checkpoint with given timestamp.
+    async fn make_checkpoint(&self, checkpoint: Checkpoint) -> anyhow::Result<()>;
+    /// Loads last seen generation from the checkpoint table.
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>>;
+    /// Loads last seen timestamp for given stream id from the checkpoint table.
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>>;
+}
+
+/// Default implementation for CDCCheckpointMaker trait.
+pub struct CDCDefaultCheckpointMaker {
+    session: Arc<Session>,
+    keyspace: String,
+    checkpoint_tablename: String,
+    prepared_stmt: PreparedStatement,
+}
+
+impl CDCDefaultCheckpointMaker {
+    /// Creates new CDCDefaultCheckpointMaker.
+    /// If given None value as table name,
+    /// it uses default value "cdc_checkpoint".
+    pub async fn new(
+        session: &Arc<Session>,
+        keyspace: String,
+        tablename: Option<String>,
+    ) -> anyhow::Result<Self> {
+        let session = Arc::clone(session);
+        let checkpoint_tablename = tablename.unwrap_or_else(|| "CDC_CHECKPOINT".to_string());
+
+        CDCDefaultCheckpointMaker::create_table(&session, &keyspace, &checkpoint_tablename).await?;
+        session.await_schema_agreement().await?;
+
+        let prepared_stmt = session
+            .prepare(format!(
+                "UPDATE {}.{} SET time = (?)
+                WHERE stream_id = (?)
+                AND generation = (?)",
+                &keyspace, &checkpoint_tablename
+            ))
+            .await?;
+        let cp_maker = CDCDefaultCheckpointMaker {
+            session,
+            keyspace,
+            checkpoint_tablename,
+            prepared_stmt,
+        };
+
+        Ok(cp_maker)
+    }
+
+    // Creates new table with specified schema in the given keyspace with provided name.
+    // If a table with such name already exists, doesn't do anything.
+    async fn create_table(
+        session: &Session,
+        keyspace: &str,
+        checkpoint_tablename: &str,
+    ) -> anyhow::Result<()> {
+        let schema = format!(
+            "CREATE TABLE IF NOT EXISTS {}.{} (
+            stream_id blob PRIMARY KEY,
+            time timestamp,
+            generation timestamp,
+            )
+            WITH CLUSTERING ORDER BY (generation DESC)",
```

Comment:
Setting "generation" as the clustering key does not make it possible to find the maximum generation with `SELECT generation FROM x LIMIT 1`.

Performing an unrestricted SELECT returns the contents of a table ordered by:

1. Partition key token; the token is a hash of the partition key, so the order looks random,
2. Clustering key, as defined by CLUSTERING ORDER.

Because of that, the rows returned by `SELECT generation FROM x` won't be necessarily ordered by generation. Besides, such queries are quite heavy (and usually slow) because they involve all nodes in the cluster, even if you use `LIMIT 1`.

What I would recommend is to have a special, single partition which holds the current generation and read/update it when needed. You can, for example, use an empty string as `stream_id` to identify the special partition.